### ISSUE TITLE
pyright --createstub cannot complete like useLibraryCodeForTypes

### ISSUE THREAD

**Issue by tshu-w** [microsoft/pyright#1970]
**Describe the bug**
`pyright --createstub` cannot enable pyright to complete 3rd party library like enable `useLibraryCodeForTypes`

**To Reproduce**
1. `pyright --createstub pandas` 
2. typed following code
3. cannot complete
4. enable useLibraryCodeForTypes
5. now pyright can complete DataFrame variable

**Expected behavior**
`pyright --createstub` behave like enable `useLibraryCodeForTypes`.

**Screenshots or Code**
```python
import pandas as pd

d = {'col1': [1, 2], 'col2': [3, 4]}
df = pd.DataFrame(data=d)
df
```

**command-line**
pyright 1.1.142

**Additional context**
If I download stubs from https://github.com/microsoft/python-type-stubs, completion is correct.

**Comment by tshu-w** [comment]
This issue also happen on some others packages like https://github.com/huggingface/transformers

**Comment by erictraut** [comment]
The "--createstub" feature is intended for purposes very different than the `useLibraryCodeForTypes`. In particular, it is designed to create a "first draft" of a type stub that you can then improve and complete (manually) for purposes of type checking. The `useLibraryCodeForTypes` is intended to create low-quality type information that's usually insufficient for type checking but may be sufficient for completion suggestions.

If your use case is to enable completion suggestions and `useLibraryCodeForTypes` is working, please continue to use that approach.

I did recently update `--createstubs` to emit as a comment the return type for functions if it can be inferred. This may reduce the manual work involved in improving the type stubs if you're creating them for the purposes of type checking.

**Comment by erictraut** [comment]
I'll also add that "--createstub" will never do a good job with a library as complicated as pandas. The pandas maintainers are working on developing inlined type information to eliminate the need for pandas stubs. In the meantime, we have created some pandas stubs that are good enough for type completion. We bundle these with pylance but not with pyright. They are available for separate download [here](https://github.com/microsoft/python-type-stubs), so you're welcome to install them for your own use.

**Comment by tshu-w** [comment]
> If your use case is to enable completion suggestions and `useLibraryCodeForTypes` is working, please continue to use that approach.

I looked for another solution because useLibraryCodeForTypes is very slow, especially with large packages like https://github.com/huggingface/transformers, so I'm wondering if there's always a way to cache something in advance to speed up the completion.

**Comment by erictraut** [comment]
Internally, pyright caches type information in memory, but it doesn't persist this cache. Pyright uses lazy evaluation to evaluate only the types that are needed for a given type query.

Are you using VS Code or some other editor. Most other editors feel much slower because they don't support file-based cancellation, so when you start typing, a completion suggestion request is issued to pyright, but if you then type another character, the first request cannot be canceled before issuing the next completion request. That could account for what you're seeing. If so, you should encourage the maintainers of the editor (or the LSP support for that editor) to implement file-based cancellation support.

**Comment by tshu-w** [comment]
@erictraut Thx for your quick reply. Yes, I'm using Emacs with pyright. There is another question. If I'm using transformers, the analyzing of py file took a long time, even if you don't enter anything.
```
[FG] Long operation: checking: file.py (9989ms)
[FG] Long operation: analyzing: file.py (9989ms)
```

**Comment by tshu-w** [comment]
@erictraut Hi, can you explain why the above checking and analyzing took a lot of time. Thx a lot.

**Comment by jakebailey** [comment]
Also related is https://github.com/microsoft/pylance-release/issues/1258.

Something about transformers is hard to analyze and I was hoping to do a little profiling tomorrow to see what.

**Comment by tshu-w** [comment]
@jakebailey Yes, I also meet issue with `AutoTokenizer.from`
